### DISCUSSION TITLE
Is persistent Heapsize memory growth a sign of memory leak?

### DISCUSSION THREAD

**Discussion by mrQanderson** [bloomberg/memray#422] [Q&A]
Hi, dear team. You are doing a great job! This product is fantastic!

@pablogsal , could you, please, kindly help me...
I was faced with the task of having a test on memory leakage. Our app (refinitiv-data) is a thin layer between User and Server.
The current script created 5000 streams by iterations. 
[testrdp.py.zip](https://github.com/bloomberg/memray/files/12053708/testrdp.py.zip)
We've got feedback about some memory leak assumptions.
_Windows perf results below:_
![Py310process](https://github.com/bloomberg/memray/assets/66297758/65fc7894-e81b-4e81-a367-c5014b3cb6b6)

Then I tested with Memray on MacOS Ventura 13.4.
_Memray 1.8.1, python 3.10_
export PYTHONMALLOC="malloc"
memray run -f -o memray_capture_develop.bin --trace-python-allocators testrdp.py
memray flamegraph --leaks memray_capture_develop.bin  
<img width="1126" alt="image" src="https://github.com/bloomberg/memray/assets/66297758/1120ee76-1227-4386-9269-58cb5a79db68">
Graph with initial issues - [memray-flamegraph-memray_capture_develop.html.zip](https://github.com/bloomberg/memray/files/12070759/memray-flamegraph-memray_capture_develop.html.zip)

Heapsize starts ~94.69Mb, ends ~113.5Mb

Do I understand correctly that Python process constantly asks OS to allocate more memory (Heapsize memory)?
Resident size memory is a real app memory that depends on allocated/deallocated objects?

Does this look like a memory leak?
How should look like an absence of memory leak, the Heapsize graph has to come back to initial(start) size ?

Also we did some app adjustments to streams
Heapsize starts ~93.75Mb, ends ~105.3Mb
<img width="1124" alt="image" src="https://github.com/bloomberg/memray/assets/66297758/cdd10286-afeb-4fdf-bb8e-5a0e008a8eb0">
Graph with fix - [memray-flamegraph-memray_capture_fix.html.zip](https://github.com/bloomberg/memray/files/12070765/memray-flamegraph-memray_capture_fix.html.zip)

Thank you for you time.

**Comment by pablogsal** [comment]
Hi, thanks for opening the discussion. Unfortunately, without looking at the application is difficult to say what's going on, and even with the application could be a challenging task. Looking at your plot it looks that this is indeed a memory leak because the heap size (what your application is requesting) is growing and growing across what seems to be regular patterns. There is some memory in these chunks that's not being deallocated.

You can try to use our new temporal flame graphs with the `--leaks` flag to inspect what's not being deallocated in each of those segments:

https://bloomberg.github.io/memray/flamegraph.html#temporal-flame-graphs
